Cleanup GitHub actions

tautschnig opened the pull request:
Please see individual commit messages.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Note down expected CI execution times

When looking to optimise CI workflows it is useful to have an idea about
the current state of affairs.

tautschnig committed:
GitHub actions: avoid trashing caches

1. We must use the same build configuration when sharing a particular
   cache.
2. Don't create additional cache entries when we can safely use one of
   the existing ones.

tautschnig committed:
GitHub actions: whitespace cleanup

Remove stray space.

tautschnig committed:
GitHub actions: unify build commands

Use single-instruction commands where possible, making the code use less
vertical space (and making different jobs easier to compare).

NlightNFotis commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -6,6 +6,7 @@ on:
       - '**'
 
 jobs:
+  # This job takes approximately 18 minutes
```
I understand that these are informational only, but how are we going to track any changes to these numbers?

Or is the plan to be more laid back and only update them if we casually observe any change?

tautschnig replied:
For now I was thinking keep this informal, but maybe we should go a little further and do actual checks? Don't know, I'm open to suggestions.

NlightNFotis commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -41,7 +41,7 @@ jobs:
       - name: Build CBMC tools
         run: |
           make -C src minisat2-download
-          make -C src CXX='ccache /usr/bin/g++' cbmc.dir goto-cc.dir goto-diff.dir -j2
```
Why did the compiler change here?

tautschnig replied:
This is to make sure we can actually use the compiler cache (and don't trash it by putting a different configuration in there).

NlightNFotis reviewed (approved):
LGTM, some minor questions.

tautschnig committed:
Run GitHub pull request checks on push to develop

This will make sure tests still pass on the merged code base. Also, this
will make sure caches are actually available on first run in pull
requests: GitHub caches are access restricted and caches created by
branches in forks will not be considered. Only cache entries for the
base branch are always available.
